crates/task-impls/src/quorum_proposal_recv/handlers.rs

```diff
@@ -0,0 +1,232 @@
+
+        // This is for the case where we form a QC but have not yet seen the previous proposal ourselves
+        let should_propose = task_state.quorum_membership.leader(new_view) == task_state.public_key
+            && high_qc.view_number == current_proposal.clone().unwrap().view_number;
+
+        if should_propose {
+            debug!(
+                "Attempting to publish proposal after voting for liveness; now in view: {}",
+                *new_view
+            );
+            broadcast_event(
+                HotShotEvent::LivenessCheckProposalRecv(proposal.data.clone()).into(),
+                event_sender,
+            )
+            .await;
+        }
+    }
+
+    // If we are missing the parent from storage, the safety check will fail.  But we can
+    // still vote if the liveness check succeeds.
+    current_proposal
+}
+
+/// Handles the `QuorumProposalRecv` event.
+pub(crate) async fn handle_quorum_proposal_recv<TYPES: NodeType, I: NodeImplementation<TYPES>>(
+    proposal: &Proposal<TYPES, QuorumProposal<TYPES>>,
+    sender: &TYPES::SignatureKey,
+    event_sender: &Sender<Arc<HotShotEvent<TYPES>>>,
+    task_state: &mut QuorumProposalRecvTaskState<TYPES, I>,
+) -> Result<Option<QuorumProposal<TYPES>>> {
+    let sender = sender.clone();
+
+    validate_proposal_view_and_certs(
+        proposal,
+        &sender,
+        task_state.cur_view,
+        &task_state.quorum_membership,
+        &task_state.timeout_membership,
+    )
+    .context("Failed to validate proposal view and attached certs")?;
+
+    let view_number = proposal.data.view_number();
+    let view_leader_key = task_state.quorum_membership.leader(view_number);
+    let justify_qc = proposal.data.justify_qc.clone();
+
+    if !justify_qc.is_valid_cert(task_state.quorum_membership.as_ref()) {
+        let consensus = task_state.consensus.read().await;
+        consensus.metrics.invalid_qc.update(1);
+        bail!("Invalid justify_qc in proposal for view {}", *view_number);
+    }
+
+    // NOTE: We could update our view with a valid TC but invalid QC, but that is not what we do here
+    if let Err(e) = update_view::<TYPES>(
+        view_number,
+        event_sender,
+        task_state.timeout,
+        Arc::clone(&task_state.consensus),
+        &mut task_state.cur_view,
+        &mut task_state.timeout_task,
+        &task_state.output_event_stream,
+        SEND_VIEW_CHANGE_EVENT,
+    )
+    .await
+    {
+        debug!("Failed to update view; error = {e:#}");
+    }
+
+    let consensus_read = task_state.consensus.upgradable_read().await;
+
+    // Get the parent leaf and state.
+    let parent = match consensus_read
+        .saved_leaves()
+        .get(&justify_qc.data.leaf_commit)
+        .cloned()
+    {
+        Some(leaf) => {
+            if let (Some(state), _) = consensus_read.state_and_delta(leaf.view_number()) {
+                Some((leaf, Arc::clone(&state)))
+            } else {
```

Comment:
Saved leaves and saved state seem to be tied together maybe we should make it impossible to set one without the other.  Not for this PR but a good follow up issue

Reply:
Tracking: https://github.com/EspressoSystems/HotShot/issues/3215